### PR TITLE
Download document with its attachments

### PR DESCRIPTION
**Pull request by Temu4**
Provide different options for `GET: /document/{documentUniqueId}/download/` method.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -216,13 +216,17 @@ More: [CLI applet](https://github.com/signnow/SignNowNodeSDK/blob/master/bin/vie
 
 #### <a name="download-document"></a>Download a Collapsed Document
 
-By default document downloads without history. To download it with history set `withHistory` option to `true`.
+There is a possibility to download a document by original (by default) or archived in zip (set `zip` option to `true`).
+Also, default document downloads without history. To download it with history set `withHistory` option to `true`.
```
Comment (russellswift) [inline README.md]:
```suggestion
```

File: README.md [hunk 2]
```diff
@@ -216,13 +216,17 @@ More: [CLI applet](https://github.com/signnow/SignNowNodeSDK/blob/master/bin/vie
 
 #### <a name="download-document"></a>Download a Collapsed Document
 
-By default document downloads without history. To download it with history set `withHistory` option to `true`.
+There is a possibility to download a document by original (by default) or archived in zip (set `zip` option to `true`).
```
Comment (russellswift) [inline README.md]:
```suggestion
By default document is downloaded without history or attachments. To download it with history set `withHistory` option to `true`. To download it with attachments set `withAttachments` option to `true`.
```

File: README.md [hunk 3]
```diff
@@ -216,13 +216,17 @@ More: [CLI applet](https://github.com/signnow/SignNowNodeSDK/blob/master/bin/vie
...
+Also, default document downloads without history. To download it with history set `withHistory` option to `true`.
 
 ```javascript
 api.document.download({
   token: 'your auth token',
   id: 'document id',
-  options: { withHistory: true }, // false by default
+  options: { 
+    zip: true // false by default
```
Comment (russellswift) [inline README.md]:
```suggestion
    withAttachments: true, // false by default
```

File: README.md [hunk 4]
```diff
@@ -216,13 +216,17 @@ More: [CLI applet](https://github.com/signnow/SignNowNodeSDK/blob/master/bin/vie
...
+    withHistory: true // false by default
```
Comment (russellswift) [inline README.md]:
```suggestion
    withHistory: true, // false by default
```

File: README.md [hunk 5]
```diff
@@ -216,13 +216,17 @@ More: [CLI applet](https://github.com/signnow/SignNowNodeSDK/blob/master/bin/vie
...
+    },
```
Comment (russellswift) [inline README.md]:
```suggestion
  },
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -10,6 +10,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 
 ### Changed
+- Added optional ability to download document archived in zip for *Download Document* feature
```
Comment (russellswift) [inline CHANGELOG.md]: see example at https://github.com/signnow/SignNowNodeSDK/pull/74/files#diff-4ac32a78649ca5bdd8e0ba38b7006a1eR12

File: bin/download-document.js [hunk 7]
```diff
@@ -2,22 +2,30 @@
 
 /**
  * to run download-document applet from the project root folder type in your console:
- * > node bin/download-document <clienе_id> <client_secret> <username> <password> <document_id> <path_to_save> <with_history>
+ * > node bin/download-document <clienе_id> <client_secret> <username> <password> <document_id> <path_to_save>
```
Comment (russellswift) [inline bin/download-document.js]:
```suggestion
 * > node bin/download-document <client_id> <client_secret> <username> <password> <document_id> <path_to_save>
```

File: bin/download-document.js [hunk 8]
```diff
@@ -2,22 +2,30 @@
...
  * <client_id>, <client_secret>, <username>, <password>, <document_id>, <path_to_save> - are required params
- * <with_history> - optional param
+ * options flags:
```
Comment (russellswift) [inline bin/download-document.js]:
```suggestion
 * options:
```

File: bin/download-document.js [hunk 9]
```diff
@@ -2,22 +2,30 @@
...
+ * --zip - document will be archived in zip
```
Comment (russellswift) [inline bin/download-document.js]:
```suggestion
 * --with-attachments - document will be downloaded as zip package with its attachments
```

File: bin/download-document.js [hunk 10]
```diff
@@ -2,22 +2,30 @@
...
   clientSecret,
   username,
   password,
   documentId,
   pathToSaveFile,
-  withHistory,
-] = process.argv.slice(2);
+] = params;
+
+const zip = flags.includes('--zip');
```
Comment (russellswift) [inline bin/download-document.js]:
```suggestion
const withAttachments = flags.includes('--with-attachments');
```

File: bin/download-document.js [hunk 11]
```diff
@@ -40,12 +48,15 @@ getAccessToken$({
 })
   .then(({ access_token: token }) => downloadDocument$({
     id: documentId,
-    options: { withHistory: withHistory === 'true' },
+    options: { 
+      zip,
```
Comment (russellswift) [inline bin/download-document.js]:
```suggestion
      withAttachments,
```

File: lib/document/index.js [hunk 12]
```diff
@@ -540,7 +540,7 @@ class Document {
   /**
    * Download document optional settings
    * @typedef {Object} DocumentDownloadOptions
-   * @property {boolean} [collapsed=true] - if true document will be downloaded with embedded field and element values
+   * @property {boolean} [zip=false] - if false document will be downloaded with embedded field and element values else archived to zip
```
Comment (russellswift) [inline lib/document/index.js]:
```suggestion
   * @property {boolean} [withAttachments=false] - if true document will be downloaded as zip package with its attachments
```

File: lib/document/index.js [hunk 13]
```diff
@@ -568,15 +568,15 @@ class Document {
     id,
     token,
     options: {
-      collapsed = true,
+      zip = false,
```
Comment (russellswift) [inline lib/document/index.js]:
```suggestion
      withAttachments = false,
```

File: lib/document/index.js [hunk 14]
```diff
@@ -568,15 +568,15 @@ class Document {
...
       withHistory = false,
     } = {},
   }, callback) {
     const queryParams = {};
     const clientTimestamp = Math.floor(Date.now() / 1000);
 
-    collapsed && (queryParams.type = 'collapsed');
-    withHistory && (queryParams.with_history = '1');
+    zip ? (queryParams.type = 'zip') : (queryParams.type = 'collapsed');
```
Comment (russellswift) [inline lib/document/index.js]:
```suggestion
    queryParams.type = withAttachments ? 'zip' : 'collapsed';
```

File: lib/document/index.js [hunk 15]
```diff
@@ -568,15 +568,15 @@ class Document {
...
+    withHistory ? (queryParams.with_history = '1') : null;
```
Comment (russellswift) [inline lib/document/index.js]:
```suggestion
    withHistory && (queryParams.with_history = '1');
```

File: samples/snippets/downloadDocumentWithAttachments.js [hunk 16]
```diff
@@ -0,0 +1,35 @@
+'use strict';
+
+const signnow = require('@signnow/api-client')({
+  credentials: 'BASE64_ENCODED_CLIENT_CREDENTIALS',
+  production: true, // if false then uses eval server
+});
+const downloadDocument = signnow.document.download;
```
Comment (russellswift) [inline samples/snippets/downloadDocumentWithAttachments.js]:
```suggestion
const { download: downloadDocument } = signnow.document;
```

File: samples/snippets/downloadDocumentWithAttachments.js [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+const id = 'DOCUMENT_ID_GOES_HERE';
+const token = 'YOUR_ACCESS_TOKEN';
+const options = { zip: true }; // option points on downloading document archived in zip
```
Comment (russellswift) [inline samples/snippets/downloadDocumentWithAttachments.js]:
```suggestion
const options = { withAttachments: true };
```

File: samples/snippets/downloadDocumentWithAttachments.js [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * @param {Buffer} res - binary data (file) of the entire document
```
Comment (russellswift) [inline samples/snippets/downloadDocumentWithAttachments.js]:
```suggestion
 * @param {Buffer} res - binary data of zip package
```

File: samples/snippets/downloadDocumentWithAttachments.js [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+ */
+
+const handleResponse = res => {
+  console.log(res); // save the file to your disk or pipe it to another handler
```
Comment (russellswift) [inline samples/snippets/downloadDocumentWithAttachments.js]:
```suggestion
  // save the file to your disk or pipe it to another handler
```

File: bin/download-document.js (hunk 10 above)
Comment (Temu4) [reply bin/download-document.js]: IMO `withAttachments` not as clear as `zip`... but if you think so, it's OK.

File: bin/download-document.js [hunk 20]
```diff
@@ -46,12 +48,15 @@ getAccessToken$({
...
     id: documentId,
-    options: { withHistory },
+    options: { 
+      withAttachments,
+      withHistory,
+     },
     token,
   }))
   .then(file => {
     const absolutePath = `${pathToSaveFile}/${documentId}.pdf`;
```
Comment (russellswift) [inline bin/download-document.js]: file may be .zip

File: bin/download-document.js [hunk 21]
```diff
@@ -46,12 +48,17 @@ getAccessToken$({
...
+    options: { 
+      withAttachments,
+      withHistory,
+     },
     token,
   }))
   .then(file => {
-    const absolutePath = `${pathToSaveFile}/${documentId}.pdf`;
+    let fileExtension = '';
+    withAttachments? fileExtension = 'zip' : fileExtension = 'pdf';
```
Comment (russellswift) [inline bin/download-document.js]:
```suggestion
```

File: bin/download-document.js [hunk 22]
```diff
@@ -46,12 +48,17 @@ getAccessToken$({
...
-    options: { withHistory },
+    options: { 
+      withAttachments,
+      withHistory,
+     },
     token,
   }))
   .then(file => {
-    const absolutePath = `${pathToSaveFile}/${documentId}.pdf`;
+    let fileExtension = '';
```
Comment (russellswift) [inline bin/download-document.js]:
```suggestion
    const fileExtension = withAttachments ? 'zip' : 'pdf';
```

**Pull request merged**